DigixGlobal/governance-ui-components
# Refactor proposal forms and require all fields except file uploads

Ref: [DGDG-433](https://tracker.digixdev.com/issue/DGDG-403)

### Dev Notes
This creates the `<ProposalForms/>` component, which serves as the base component for `<CreateProposal/>` and `<EditProposal/>` and keeps the code DRY. All fields when creating or modifying projects are now required, with the exception for file uploads.

Please take note of the following changes:
- `<Preview/>` has been removed from `proposals/create` and `proposals/edit` and instead moved to the `proposals/form` directory. `<ProposalForms/>` and its styles (formerly in `proposals/create`) has been moved here as well. The styles file has been renamed to `base-styles`.
- `<ProposalForms/>` will handle the form validation. The individual forms will render the errors based on the `errors` props passed to them.
- The tab panel is no longer clickable since we're disabling the `Next` button whenever the current form is invalid. The user should no longer be able to jump between forms to ensure that they fill up all required fields.
- A fix has been added for a bug introduced in PR #172. The current implementation assumes the max number of milestones to slice is two (2); the fix ensures it should work for any milestone cap. Moreover, it replaces the use of `Array#splice` with `Array#slice` so that it behaves correctly with React, since `Array#slice` is immutable. For more info, see: https://stackoverflow.com/a/50579752.

### Code Review Notes

Note that `<ProposalForms/>` (found in `src/pages/proposals/forms/index.js`) will not be initially visible in this PR.  You have to manually load the diff since it's a bit too large. Aside from the field validations, the logic is mostly the same from `<CreateProposal/>` and `<EditProposal/`>.

### Test Plan
- Create a proposal. The form should require all fields except file uploads.
- Edit a project. The form should require all fields except file uploads.
- Regression tests for creating and editing projects should pass.
